### ISSUE TITLE
DAP UI Plugin

### ISSUE THREAD

**Issue by kobzar** [LunarVim/LunarVim#2780]
### Feature Description

We are use DAP and DAPInstall as core plugins. 
But clear DAP without UI is bad. 
So, if we use it in core config - make it better by default with
https://github.com/rcarriga/nvim-dap-ui
Thats LunarVim debugger will look normaly like in others editors/ide

Yes, we can add it manualy in plugin list - but i think this is part of lvim phylosophy as IDE

**Comment by rebuilt** [comment]
I'm surprised to say I agree with the idea that debugging is in an akward state.  Including dap as one of the core plugins makes it  seem like it's a supported feature.  I know you can turn on dap with this command:
```
lvim.builtin.dap.active = true
```
But then what do I do?  Our documentation is very sparse.  There isn't a section that explains how to set everything up correctly.  Some of the language specific sections have dap configurations.  Go is an example of a language that does include dap configurations  https://www.lunarvim.org/languages/go.html#debugger. But for a lot of languages, the user is on their own.

Compare that with the LSP story in LunarVim:  We've done an amazing job getting people up and running with LSP.  It's easy to do in LunarVim because we've added infrastructure to make setup easy (Thanks @kylo252)  And the LunarVim documentation has all the steps listed to give users a quick start.   https://www.lunarvim.org/languages/#server-setup

We have some options.  I'm not sure what the best(and they're not mutually exclusive).  But I'll just list options to get the conversation going.
We can leave things as they are.
We could add documentation.
We could add functionality to LunarVim to streamline language configurations for DAP as we have for LSP.
We can add the dap ui plugin and make sure it plays well with dap and dap-buddy.

We should keep in mind that adding these features will mean we'll end up supporting debugger problems when it breaks.  @abzcoding Your opinion on this will add a lot of weight.  

In the mean time, I'll set up debugging for myself and see how much I can contribute to the documentation.

**Comment by bestouff** [comment]
Moreover the debugger installer used, [`dap-buddy`](Pocco81/dap-buddy.nvim) plainly doesn't work and has no documentation (it's apparently being rewritten currently).
For the installer part, maybe replacing `dap-buddy` + `nvim-lsp-installer` by [`mason`](williamboman/mason.nvim) would be an option ? But that's a breaking change ...

**Comment by bestouff** [comment]
Oh yes the DAP experience isn't up to the rest of lvim. I wish it worked out of the box.

**Comment by longguzzz** [comment]
Here is the series of operations for someone new who want to try it.

**Step 1, follow the installation steps in [Debug Adapter installation · mfussenegger/nvim-dap Wiki](https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#Go)**
Install [delve](https://github.com/go-delve/delve/tree/master/Documentation/installation)
```
go install github.com/go-delve/delve/cmd/dlv@latest
```
or via package manager (`pacman -S delve`)

Here, change current directory to your user directory.
And install [vscode-go](https://github.com/golang/vscode-go)
```
git clone https://github.com/golang/vscode-go
cd vscode-go
npm install
npm run compile
```
The path of debugAdapter.js in `vscode-go\dist` will be used later. 

**Step 2, modify your config.lua**
Add the code below to your config.lua
```
lvim.builtin.dap.active = true
local dap = require "dap"
dap.adapters.go = {
  type = 'executable';
  command = 'node';
  args = { <you_debugAdapter_path>  }; -- specify the path to the adapter
}
dap.configurations.go = {
  {
    type = "go",
    name = "Attach",
    request = "attach",
    processId = require("dap.utils").pick_process,
    program = "${workspaceFolder}",
    dlvToolPath = vim.fn.exepath('dlv')
  },
  {
    type = "go",
    name = "Debug curr file",
    request = "launch",
    program = "${file}",
    dlvToolPath = vim.fn.exepath('dlv')
  },
  {
    type = "go",
    name = "Debug",
    request = "launch",
    program = "${workspaceFolder}",
    dlvToolPath = vim.fn.exepath('dlv')
  },
  {
    type = "go",
    name = "Debug curr test",
    request = "launch",
    mode = "test",
    program = "${file}",
    dlvToolPath = vim.fn.exepath('dlv')
  },
  {
    type = "go",
    name = "Debug test",
    request = "launch",
    mode = "test",
    program = "${workspaceFolder}",
    dlvToolPath = vim.fn.exepath('dlv')
  },
}
```
In windows, config.lua is in `~/AppData/Local/lvim`. Or you can check `:h base-directories`, which is similar.
In windows, replace `<you_debugAdapter_path>` to `C:/Users/<yourname>/vscode-go/dist/debugAdapter.js`. `<yourname>` is your user name. Take attention that `os.getenv('HOME')` in [Go | LunarVim Docs](https://www.lunarvim.org/languages/go.html#install-syntax-highlighting) may return nil, so you may need to type full path yourself.

Next, modify `lvim.plugins` in the same file with adding  ` { "rcarriga/nvim-dap-ui", requires = {"mfussenegger/nvim-dap"} },` in  like this.
```
lvim.plugins = {
    { "rcarriga/nvim-dap-ui", requires = {"mfussenegger/nvim-dap"} },
}
```


**Step 3, try to use**
1. Open your lvim and `:PackerSync`.
2. After installation, Open a simple go file.
3. For [rcarriga/nvim-dap-ui](https://github.com/rcarriga/nvim-dap-ui), try `:lua require("dapui").setup()` and `:lua require("dapui").open()` in some go file.
4. Try to press `<leader>d` and wait the prompt of which-key.
The rest will be easy.

**Comment by bestouff** [comment]
Thanks. Not exactly user-friendly ...

**Comment by amulepeweichan** [comment]
Instead of Step 1, I was able to install Delve and the debuggers for 
other languages I use (rust, C, python) from LunarVim from the included 
Mason package manager, which was nice. But yea the rest wasn't so easy 
to figure out.

**Comment by albertas** [comment]
I would love to get step-by-step guide for Python DAP setup for LunarVim.

**Comment by amulepeweichan** [comment]
In Mason (space-l-I to open) I selected to install debugpy, then in the 
config to lvim.plugins I added:
```
{ "mfussenegger/nvim-dap-python" },
```
And in `.config/lvim/ftplugin/python.lua` I added:
```
require("dap-python").setup("~/.local/share/nvim/mason/packages/debugpy/venv/bin/python")
```
Takes care of the python specific config. For the general DAP config I 
didn't install the whole `nvim-dap-ui`, just default nvim-dap seems to 
come with shortcuts if you do space-d you see which-key menu to start 
debugger, set breakpoints, continue, step etc. And I added plugins 
`theHamsta/nvim-dap-virtual-text` which shows values of your local 
variables in text next to them in the code, and 
`nvim-telescope/telescope-dap.nvim` which adds telescope menu to search 
all variables, stackframes, breakpoints and something else I forget

**Comment by lvimuser** [comment]
example with platform independent paths
```lua
local join_paths = require("lvim.utils").join_paths
local function get_install_path(package_name)
  return require("mason-registry").get_package(package_name):get_install_path()
end

join_paths(get_install_path "debugpy", "debugpy-adapter"),
```

or

```lua
join_paths(vim.fn.stdpath("data"), "/mason/packages/debugpy/debugpy-adapter")
```

**Comment by kylo252** [comment]
fixed in #3127

**Comment by yonydev** [comment]
Hi Does anyone know how to configure it for typescript ?
